alf/algorithms/sarsa_algorithm.py

```diff
@@ -90,179 +210,343 @@ def __init__(self,
+        target_models = copy.copy(self._target_critic_networks)
+
+        self._rollout_actor_network = self._actor_network
+        if use_smoothed_actor:
+            assert not on_policy, ("use_smoothed_actor can only be used in "
+                                   "off-policy training")
+            self._rollout_actor_network = actor_network.copy(
+                name='rollout_actor_network')
+            models.append(self._actor_network)
+            target_models.append(self._rollout_actor_network)
+
         self._update_target = common.get_target_updater(
-            models=[self._actor_network, self._critic_network],
-            target_models=[
-                self._target_actor_network, self._target_critic_network
-            ],
+            models=models,
+            target_models=target_models,
             tau=target_update_tau,
             period=target_update_period)
+
         self._dqda_clipping = dqda_clipping
-        self._gamma = gamma
-        self._ou_process = create_ou_process(action_spec, ou_stddev,
-                                             ou_damping)
+
+        # TODO: make a batched OUProcess. The current one is not batched.
+        # Different environments use the same OU process, which is not good.
+        self._ou_process = common.create_ou_process(action_spec, ou_stddev,
+                                                    ou_damping)
+        self._critic_losses = []
+        for i in range(num_replicas):
+            self._critic_losses.append(
+                critic_loss_cls(debug_summaries=debug_summaries and i == 0))
+
+        self._critic_bias = None
+        if fast_critic_bias_speed >= 1:
+            self._critic_bias = FastCriticBias(
+                num_critics=num_replicas, speed=fast_critic_bias_speed)
+
+        self._is_rnn = len(alf.nest.flatten(critic_network.state_spec)) > 0
+
+    def is_on_policy(self):
+        return self._on_policy
 
     def _trainable_attributes_to_ignore(self):
-        return ["_target_actor_network", "_target_critic_network"]
+        return ["_target_critic_networks", "_rollout_actor_network"]
 
     def _get_action(self, actor_network, time_step, state, epsilon_greedy=1.0):
         action_distribution, state = actor_network(
-            time_step.observation,
-            step_type=time_step.step_type,
-            network_state=state)
+            time_step.observation, state=state)
         if isinstance(actor_network, DistributionNetwork):
-            action = common.epsilon_greedy_sample(action_distribution,
-                                                  epsilon_greedy)
+            if epsilon_greedy == 1.0:
+                action = dist_utils.rsample_action_distribution(
+                    action_distribution)
+            else:
+                action = dist_utils.epsilon_greedy_sample(
+                    action_distribution, epsilon_greedy)
         else:
 
             def _sample(a, ou):
```

Comment:
This is a questions shared with ddpg. The ```_sample``` function in tf version performs epsilon-greedy operation at the batch-level. In your pytorch implementation, this is operated as the sample-level. Do you think we should all switch to sample-level as you done here? Not sure if there is a clear difference empirically.

Reply:
tf version is also batch level

Reply:
I might didn't make the point clear. For the ```_sample``` function, it seems:

TF (Line 161-164 of tf code) performs epsilon-greedy operation at batch-level
Pytorch (Line 341-346 of pytorch code): epsilon-greedy operation at sample-level.

Did I interpret the code incorrectly?